cmd/tk/create_source.go

```diff
@@ -0,0 +1,236 @@
+	if err != nil {
+		return fmt.Errorf("git URL parse failed: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	withAuth := false
+	if strings.HasPrefix(sourceGitURL, "ssh") {
+		if err := generateSSH(name, u.Host, tmpDir); err != nil {
+			return err
+		}
+		withAuth = true
+	} else if sourceUsername != "" && sourcePassword != "" {
+		if err := generateBasicAuth(name); err != nil {
+			return err
+		}
+		withAuth = true
+	}
+
+	logAction("generating source %s in %s namespace", name, namespace)
+
+	t, err := template.New("tmpl").Parse(gitSource)
+	if err != nil {
+		return fmt.Errorf("template parse error: %w", err)
+	}
+
+	source := struct {
+		Name      string
+		Namespace string
+		URL       string
+		Branch    string
+		Semver    string
+		Interval  string
+		WithAuth  bool
+	}{
+		Name:      name,
+		Namespace: namespace,
+		URL:       sourceGitURL,
+		Branch:    sourceGitBranch,
+		Semver:    sourceGitSemver,
+		Interval:  interval,
+		WithAuth:  withAuth,
+	}
+
+	var data bytes.Buffer
+	writer := bufio.NewWriter(&data)
+	if err := t.Execute(writer, source); err != nil {
+		return fmt.Errorf("template execution failed: %w", err)
+	}
+	if err := writer.Flush(); err != nil {
+		return fmt.Errorf("source flush failed: %w", err)
+	}
+
+	if sourceVerbose {
+		fmt.Print(data.String())
+	}
+
+	command := fmt.Sprintf("echo '%s' | kubectl apply -f-", data.String())
+	c := exec.CommandContext(ctx, "/bin/sh", "-c", command)
+
+	var stdoutBuf, stderrBuf bytes.Buffer
+	c.Stdout = io.MultiWriter(os.Stdout, &stdoutBuf)
+	c.Stderr = io.MultiWriter(os.Stderr, &stderrBuf)
+
+	err = c.Run()
+	if err != nil {
+		return fmt.Errorf("source apply failed")
+	}
+
+	logAction("waiting for source sync")
+	if output, err := execCommand(fmt.Sprintf(
+		"kubectl -n %s wait gitrepository/%s --for=condition=ready --timeout=1m",
+		namespace, name)); err != nil {
+		return fmt.Errorf("source sync failed: %s", output)
+	} else {
+		fmt.Print(output)
+	}
+
+	return nil
+}
+
+func generateBasicAuth(name string) error {
+	logAction("saving credentials")
+	credentials := fmt.Sprintf("--from-literal=username='%s' --from-literal=password='%s'",
+		sourceUsername, sourcePassword)
+	secret := fmt.Sprintf("kubectl -n %s create secret generic %s %s --dry-run=client -oyaml | kubectl apply -f-",
+		namespace, name, credentials)
+	if output, err := execCommand(secret); err != nil {
+		return fmt.Errorf("kubectl create secret failed: %s", output)
+	} else {
+		fmt.Print(output)
+	}
+	return nil
+}
+
+func generateSSH(name, host, tmpDir string) error {
+	logAction("generating host key for %s", host)
+
+	keyscan := fmt.Sprintf("ssh-keyscan %s > %s/known_hosts", host, tmpDir)
+	if output, err := execCommand(keyscan); err != nil {
+		return fmt.Errorf("ssh-keyscan failed: %s", output)
+	}
+
+	logAction("generating deploy key")
+
+	keygen := fmt.Sprintf("ssh-keygen -b 2048 -t rsa -f %s/identity -q -N \"\"", tmpDir)
```

Comment:
This could be done without depending on `ssh-keygen` using one of the `crypto` packages.

Reply:
If we want to get rid of openssh we need to replace keyscan as well. We could generate the host keys with something like https://github.com/Eun/sshkeys/blob/master/sshkeys.go